### ISSUE TITLE
Help wanted with Docker

### ISSUE THREAD

**Issue by commagray** [renatolond/mastodon-twitter-poster#306]
Hi!

I'm trying to containerize this thing at https://moonbutt.science/innereq/containers/crossposter, but I'm getting Internal Server Error while logging into Twitter or Mastodon.

Could someone test it and tell me what have I done wrong?

Thanks.

![image](https://user-images.githubusercontent.com/27947961/72684110-60414200-3ae6-11ea-9ed1-cb0da5b2811e.png)

```
nightmare_1  | Puma starting in single mode...
nightmare_1  | * Version 4.3.1 (ruby 2.5.7-p206), codename: Mysterious Traveller
nightmare_1  | * Min threads: 5, max threads: 5
nightmare_1  | * Environment: production
nightmare_1  | * Listening on tcp://0.0.0.0:3001
nightmare_1  | Use Ctrl-C to stop
db_1         | 
db_1         | PostgreSQL Database directory appears to contain a database; Skipping initialization
db_1         | 
db_1         | 2020-01-19 16:03:56.992 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1         | 2020-01-19 16:03:56.992 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1         | 2020-01-19 16:03:56.994 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1         | 2020-01-19 16:03:57.030 UTC [20] LOG:  database system was shut down at 2020-01-19 14:50:52 UTC
db_1         | 2020-01-19 16:03:57.036 UTC [1] LOG:  database system is ready to accept connections
redis_1      | 1:C 19 Jan 2020 16:03:56.949 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
redis_1      | 1:C 19 Jan 2020 16:03:56.949 # Redis version=5.0.7, bits=64, commit=00000000, modified=0, pid=1, just started
redis_1      | 1:C 19 Jan 2020 16:03:56.949 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
redis_1      | 1:M 19 Jan 2020 16:03:56.952 * Running mode=standalone, port=6379.
redis_1      | 1:M 19 Jan 2020 16:03:56.952 # Server initialized
redis_1      | 1:M 19 Jan 2020 16:03:56.952 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
redis_1      | 1:M 19 Jan 2020 16:03:56.952 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
redis_1      | 1:M 19 Jan 2020 16:03:56.953 * DB loaded from disk: 0.000 seconds
redis_1      | 1:M 19 Jan 2020 16:03:56.953 * Ready to accept connections
nightmare_1  | [8969d361-5ad6-4347-b197-b8090b456e69] Started GET "/" for 172.18.0.3 at 2020-01-19 16:04:16 +0000
nightmare_1  | [8969d361-5ad6-4347-b197-b8090b456e69] Processing by HomeController#index as HTML
nightmare_1  | [8969d361-5ad6-4347-b197-b8090b456e69]   Rendering home/index.html.erb within layouts/application
nightmare_1  | [8969d361-5ad6-4347-b197-b8090b456e69]   Rendered home/index.html.erb within layouts/application (1124.5ms)
nightmare_1  | [8969d361-5ad6-4347-b197-b8090b456e69] Completed 200 OK in 1139ms (Views: 1136.8ms)
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] Started POST "/users/auth/twitter" for 172.18.0.3 at 2020-01-19 16:04:19 +0000
nightmare_1  | I, [2020-01-19T16:04:19.673138 #6]  INFO -- omniauth: (twitter) Request phase initiated.
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8]   
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] OAuth::Unauthorized (401 Authorization Required):
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8]   
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/oauth-0.5.4/lib/oauth/consumer.rb:236:in `token_request'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/oauth-0.5.4/lib/oauth/consumer.rb:155:in `get_request_token'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/omniauth-oauth-1.1.0/lib/omniauth/strategies/oauth.rb:28:in `request_phase'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/omniauth-twitter-1.4.0/lib/omniauth/strategies/twitter.rb:61:in `request_phase'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:226:in `request_call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:188:in `call!'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:169:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/http_accept_language-2.1.1/lib/http_accept_language/middleware.rb:14:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/warden-1.2.8/lib/warden/manager.rb:36:in `block in call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/warden-1.2.8/lib/warden/manager.rb:34:in `catch'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/warden-1.2.8/lib/warden/manager.rb:34:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/tempfile_reaper.rb:15:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/etag.rb:25:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/conditional_get.rb:38:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/head.rb:12:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/http/content_security_policy.rb:18:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/session/abstract/id.rb:259:in `context'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/session/abstract/id.rb:253:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/cookies.rb:670:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.4.1/lib/active_support/callbacks.rb:98:in `run_callbacks'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.1/lib/rails/rack/logger.rb:38:in `call_app'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.1/lib/rails/rack/logger.rb:26:in `block in call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.4.1/lib/active_support/tagged_logging.rb:71:in `block in tagged'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.4.1/lib/active_support/tagged_logging.rb:28:in `tagged'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.4.1/lib/active_support/tagged_logging.rb:71:in `tagged'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.1/lib/rails/rack/logger.rb:26:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/request_id.rb:27:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/method_override.rb:22:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/runtime.rb:22:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.4.1/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/executor.rb:14:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/actionpack-5.2.4.1/lib/action_dispatch/middleware/static.rb:127:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/rack-2.0.8/lib/rack/sendfile.rb:111:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.1/lib/rails/engine.rb:524:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/puma-4.3.1/lib/puma/configuration.rb:228:in `call'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/puma-4.3.1/lib/puma/server.rb:681:in `handle_request'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/puma-4.3.1/lib/puma/server.rb:472:in `process_client'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/puma-4.3.1/lib/puma/server.rb:328:in `block in run'
nightmare_1  | [3723d2b2-b3c3-47cd-82de-6fde2e4bbca8] vendor/bundle/ruby/2.5.0/gems/puma-4.3.1/lib/puma/thread_pool.rb:134:in `block in spawn_thread'
```

Mastodon returns this:

```
nightmare_1  | [5edde38f-a390-4a46-980b-3ebe560fd91a] Completed 500 Internal Server Error in 82ms (ActiveRecord: 12.8ms)
nightmare_1  | [5edde38f-a390-4a46-980b-3ebe560fd91a]   
nightmare_1  | [5edde38f-a390-4a46-980b-3ebe560fd91a] NameError (undefined local variable or method `uid' for #<AuthorizationCreation:0x00007fa17c432ae0>):
```

**Comment by renatolond** [comment]
Hey!
I will look into this with more detail later, but I can't seem to understand when in the Mastodon flow the case you describe would happen.
For twitter make sure the address of your server matches the one in twitter configuration, the callback url should be the same or twitter will give unauthorized as a result.

**Comment by commagray** [comment]
Ahh. Maybe I need help with this. Here is a Twitter app configuration:

![image](https://user-images.githubusercontent.com/27947961/72688169-541cab00-3b0d-11ea-89eb-24e3814a02ad.png)

API keys are correct, because without them crossposter wont work.

**Comment by renatolond** [comment]
It should be `<your_url>/users/auth/twitter/callback ` :)

**Comment by renatolond** [comment]
The callback url is the only thing should be changed, It should be `<your_url>/users/auth/twitter/callback ` :)

**Comment by commagray** [comment]
Yay! It works. Thanks. Now we only need to fix #307.

**Comment by commagray** [comment]
Yay! It works. Thanks. Now we only need to fix #307.

**Comment by commagray** [comment]
Okay, I figured it out. My instance uses WEB_DOMAIN, so both ministry.moonbutt.science and moonbutt.science can be used while login. Relay broken while using moonbutt.science, but works with ministry.moonbutt.science. Is this a bug?

My issue is solved, I think.

**Comment by renatolond** [comment]
It's most likely a bug, I don't think I've ever tested around instances with a different `web_domain`, now the error seems to make sense. I'll check it out when I have some time, thanks for reporting!

**Comment by renatolond** [comment]
Hey, I was checking this out and it seems to me that your config for the WEB_DOMAIN/LOCAL_DOMAIN seems to be different of what used to be the recommended config. I can't seem to find an up-to-date doc about it, but this is what I used at the time I configured my own server: https://github.com/ThibG/documentation/blob/0a4b6de9fa72416ac6452a81621fb056be0053f1/Running-Mastodon/Serving_a_different_domain.md

With your setup what seems to happen is that you are serving mastodon on two different domains, when the crossposter tries the web_domain, it will get back your user, but the domain will not be the same as the one it registered in the initial oauth step and it will fail.
I will try to make the error more clear, but I can't fix the issue without a much deeper change in the crossposter.
